src/libostree/ostree-repo-finder.h

```diff
@@ -0,0 +1,116 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors:
+ *  - Philip Withnall <withnall@endlessm.com>
+ */
+
+#pragma once
+
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib-object.h>
+
+#include "ostree-types.h"
+
+G_BEGIN_DECLS
+
+#define OSTREE_TYPE_REPO_FINDER (ostree_repo_finder_get_type ())
+
+_OSTREE_PUBLIC
+G_DECLARE_INTERFACE (OstreeRepoFinder, ostree_repo_finder, OSTREE, REPO_FINDER, GObject)
+
+struct _OstreeRepoFinderInterface
+{
+  GTypeInterface g_iface;
+
+  void (*resolve_async) (OstreeRepoFinder    *self,
+                         const gchar * const *refs,
+                         GCancellable        *cancellable,
+                         GAsyncReadyCallback  callback,
+                         gpointer             user_data);
+  GPtrArray *(*resolve_finish) (OstreeRepoFinder  *self,
+                                GAsyncResult      *result,
+                                GError           **error);
+};
+
+_OSTREE_PUBLIC
+void ostree_repo_finder_resolve_async (OstreeRepoFinder    *self,
+                                       const gchar * const *refs,
+                                       GCancellable        *cancellable,
+                                       GAsyncReadyCallback  callback,
+                                       gpointer             user_data);
+_OSTREE_PUBLIC
+GPtrArray *ostree_repo_finder_resolve_finish (OstreeRepoFinder  *self,
+                                              GAsyncResult      *result,
+                                              GError           **error);
+
+_OSTREE_PUBLIC
+void ostree_repo_finder_resolve_all_async (OstreeRepoFinder * const *finders,
+                                           const gchar * const      *refs,
+                                           GCancellable             *cancellable,
+                                           GAsyncReadyCallback       callback,
+                                           gpointer                  user_data);
+_OSTREE_PUBLIC
+GPtrArray *ostree_repo_finder_resolve_all_finish (GAsyncResult  *result,
+                                                  GError       **error);
+
+/**
+ * OstreeRepoFinderResult:
+ * @uri: TODO
+ * @priority: TODO
+ *
+ * TODO: basically a structure which says ‘you can download these refs from
+ * this URI, and they will be up to date’, along with some relevant metadata.
+ *
+ * Since: 2017.6
+ */
+typedef struct
+{
+  /* TODO: Should I extend this to also support mirrors and metalink? Maybe add a GVariant *metadata too?
+   * For example, should a result be able to force GPG verification? */
+  /* TODO: would it support torrenting? */
+  gchar *uri;
```

Comment:
> Synthesizing a remote internally would be good. Right now it's remotes that specify GPG for example (and in some cases which key). So perhaps this API should take a remote as input and use its GPG config?

I think we’re agreed on synthesising a remote and doing `s/gchar *uri/gchar *remote_name/` then. 👍

> I guess this is where things get interesting - in the case where host-ostree+flatpak are in /ostree/repo (your case AIUI) - do we expect to update both in one "transaction"?

You are correct in your understanding — in Endless OS, both OS and flatpak objects are stored in `/ostree/repo`, but they come from several remotes (4, at the current count).

I would expect both to be pulled in one transaction, iff the set of ref names passed to `ostree_repo_find_remotes()` includes OS and flatpak refs. Deploying those refs is a separate issue, and one which I would not necessarily expect to be atomic (although that would be interesting).

I don’t think this is a requirement which changes the implementation: if we were to limit to updating refs from the OS separately from flatpak refs, for example, the code should still support pulling from multiple remote URIs if it found multiple peers on the LAN. Pulling from multiple remote URIs for multiple refs is not more complex, as far as I can see.

(Technically we could limit to pulling from a single remote URI per ref, and that would be a simplification; but at the cost of not being able to use a single transaction to pull ref sets A+B if set A and set B are only available on separate LAN peers. I’d rather not have that limitation.)

> Currently, the refname approach allows it, but I think it'd be easier to implement if the "find process" was per remote. We don't expect people to have a lot of those in general. (I.e. we're not doing this per app).

I think we might be talking cross-purposes again. `ostree_repo_find_remotes()` can’t be per-remote, since at the time it’s called, the possible remotes (including LAN peers and mount points) haven’t been enumerated. Are you using ‘remote’ here to mean ‘a particular repository and set of ref names which are published somewhere by a vendor’ rather than ‘a URI to pull from’?

> What I was thinking about when I made the original comment is - what about the case where we have a predefined remote for the host-ostree and it's all that's available? Does the "find process" return the empty set, then the app (eos-updater) needs to fall back to using it? Or is the existing remote integrated into it?

As the branch currently stands, `ostree_repo_find_remotes()` will return the empty set, but that’s not the intention. I think clients shouldn’t have to do any fallbacks from `ostree_repo_find_remotes()`.

My original intention was that `OstreeRepoFinderConfig` would supersede the `[remote "blah"]` groups in `/ostree/repo/config` with `/etc/ostree/refs.d/$ref_name` files instead, but I admit I haven’t actually worked out a transition path for it. I suspect we might need an additional `OstreeRepoFinderOldConfig` (strawman name) which returns results from the `OstreeRepo`’s `config` file. It would return all the configured remotes from there, then the code in `ostree_repo_find_remotes()` which downloads the summary files and checks them would filter out the refs which aren’t in each remote (i.e. it would assign the OS refs to the OS remote and the flatpak refs to the flatpak remote). That would mean downloading more `summary` files than `ostree_repo_pull()` currently does (one per remote, rather than one overall), but this shouldn’t really be a problem given that they’re cached locally.